contracts/FlowIDTableStaking.cdc

```diff
@@ -894,10 +898,19 @@ pub contract FlowIDTableStaking {
         /// Removes nodes by setting their weight to zero and refunding
         /// staked and delegated tokens.
         pub fun removeAndRefundNodeRecord(_ nodeID: String) {
+
             let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
 
             emit NodeRemovedAndRefunded(nodeID: nodeRecord.id, amount: nodeRecord.tokensCommitted.balance + nodeRecord.tokensStaked.balance)
 
+            // remove the refunded node from the approve list
+            // We only do this if the approve list returns a non-nil value
+            // to avoid the infinite loop with `setApprovedList`
```

Comment:
I don't really understand how this prevents the recursive call back to `removeAndRefundNodeRecord` from `setApprovedList`...

In general `/storage/idTableApproveList` is not empty, so `getApprovedList` always returns a non-nil value. I'm not understanding how this relates to the recursive call to `removeAndRefundNodeRecord` 🤔

Reply:
in `setApproveList()`, we load the approve list from storage, which clears the storage, so `getApproveList` would return `nil` since there is nothing there.

Reply:
I see. But we still execute `removeAndRefundNodeRecord` twice - are we sure that is OK?

An example this being problematic is, we emit `NodeRemovedAndRefunded` twice when `removeAndRefundNodeRecord` is the entrypoint:
- `removeAndRefundNodeRecord` is called
- `NodeRemovedAndRefunded` is emitted (1)
- `setApprovedList` is called
- `removeAndRefundNodeRecord` is recursively called from `setApprovedList`
- `NodeRemovedAndRefunded` is emitted (2)
- since `getApprovedList` returns `nil` we avoid calling `setApprovedList` again, but still execute the rest of `removeAndRefundNodeRecord` 

In general, relying on the behaviour and ordering of operations in `setApprovedList` seems brittle to me. 

What about splitting the functions into smaller pieces to avoid the two-way coupling? 
```cadence
pub fun removeAndRefundNodeRecord(...) {
  let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
  self.unsafeSetApprovedList(...)
  self.unsafeRemoveAndRefundNodeRecord(nodeRecord)
}

// sets the approved list without validating it (requires caller to validate)
priv fun unsafeSetApprovedList(...) {}

// removes and refunds the node record without also removing them from the approved-list.
priv fun unsafeRemoveAndRefundNodeRecord() {}

// can call unsafeRemoveAndRefundNodeRecord rather than removeAndRefundNodeRecord
// then use unsafeSetApprovedList
pub fun setApprovedList(...) {}
```

Reply:
Good suggestion! I did that in my latest commit. I also reorganized the position of some methods in the Admin object